### DISCUSSION TITLE
Private NPM registry lookup with self hosted renovate works only with 1st project

### DISCUSSION THREAD

**Discussion by kovasaurus** [renovatebot/renovate#18052] [Request Help] [answered]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

32.192

### If you're self-hosting Renovate, select which platform you are using.

GitLab self-hosted

### If you're self-hosting Renovate, tell us what version of the platform you run.

14.10.3-ee

### Was this something which used to work for you, and then stopped?

I never saw this working

### Describe the bug

Running renovate bot with multiple projects breaks lookup on private npm registry after first project lookup is completed.

In bot config file have a similar config like this:
```
repositories: [
   'group1/project1',
],
npmrc: `registry=${process.env.CI_NPM_REGISTRY}\n${process.env.CI_NPM_REGISTRY_AUTH}`
```
where CI_NPM_REGISTRY_AUTH is ` //private.repo.url/repository/npm/:authToken=xxxx`

Having the above config, I can successfully fetch dependencies from the private registry. Still, if the 2nd project is added under repositories (they access the same registry with the same credentials), I constantly get 401 for all dependency lookups on the provided repo for the 2nd (and subsequent projects). In contrast, the 1st project can connect and fetch it successfully.

I can switch their places inside the config.repositories, and then the one who failed will successfully fetch dependencies from the private registry, and vice versa.
After each defined project lookup, it seems like .npmrc gets nuked but not recreated for new project lookup, hence receiving 401 from private registry

Also, I have tried multiple ways to connect to the private registry mentioned in the documentation (linked [here](https://docs.renovatebot.com/getting-started/private-packages/#npm), and only then one with providing the entire npmrc in bot's config worked.
If I have a bot config like the one below, I just get 401 for all projects trying to access the registry.

```
encrypted: {
  npmToken: "token.encryped.from.app.renovatebot.com/encrypt"
},
npmrc: "registry=private.repo.url/repository/npm/\n//private.repo.url/repository/npm/:authToken=${process.env.NPM_TOKEN}
```

Also tried combining hostRules with npmrc which I found in one of the issues on this repo, but that also made all projects get 401 when trying to connect to registry.

```
hostRules: [
        {
            hostType: "npm",
            matchHost: "private.repo.url/repository/npm/",
            token: process.env.CI_NPM_REGISTRY_TOKEN
        }
    ],
npmrc: "registry=https://private.repo.url/repository/npm/",
```

Repro steps: 
1. add single repository to bot config file
2. add npmrc string pointing to private npm registry into the bot config file
3. run the job to confirm its connection to the private npm registry (by checking dependency lookup from logs)
4. add the additional repository to the bot config file (which access the same registry(
5. run the job with DEBUG logs to inspect lookups
6. the first project in the repositories list will have 200 responses from the private registry
7. second and subsequent projects in the repositories list will have 401 responses from the private registry

### Relevant debug logs

<details><summary>Lookup for 1st repo under private npm registry (success)</summary>

```
DEBUG: http statistics (repository=group1/project1)
       "urls": {
         "https://private.repo.url/repository/npm/@babel%2Fcore (GET,200)": 1,
```

</details>

<details><summary>Lookup for 2st repo under private npm registry (fail)</summary>

```
DEBUG: http statistics (repository=group2/project2)
       "urls": {
         "https://private.repo.url/repository/npm/@babel%2Fcore (GET,401)": 1,
```

</details>

### Have you created a minimal reproduction repository?

No reproduction repository

**Comment by rarkins** [comment]
Can you grep your (debug) logs for any mentions of the string `npmrc`?

**Comment by rarkins** [comment]
`matchHost: "private.repo.url/repository/npm/",` is not valid config. It must be either a hostname or a fully qualified URL starting with `https://`

**Comment by kovasaurus** [comment]
> `matchHost: "private.repo.url/repository/npm/",` is not valid config. It must be either a hostname or a fully qualified URL starting with `https://`

It was an error on my side when hiding the real URL, originally it was with `https://`


## npmrc string in the config file

Npmrc can be seen in printed out file config and combined config as well
```
DEBUG: File config
...
"hostRules": [
           {
             "hostType": "docker",
             "matchHost": "https://private.docker.registry:port",
             "username": "username",
             "password": "***********"
           }
         ],
"npmrc": "***********"
```

From 3 target projects, only sets the npmrc file for 1st on the list according to logs. That project can successfully fetch packages from a private registry.

`
DEBUG: Setting npmrc (repository=group1/project1)
`

According to logs, the bot still tries to fetch dependencies for other projects from my private npm registry, it just gets 401 (Unlike the next example where the bot will ignore my private registry and go to fetch them from the public npm registry).

## npm host rule - regular token via git variable
When I try hostRules, I can see that host rule in the file config debug log, but its missing the token property completely (not masked like password in docker rule)
```
DEBUG: File config
...
"hostRules": [
           {
             "hostType": "docker",
             "matchHost": "https://private.docker.registry:port",
             "username": "username",
             "password": "***********"
           },
           {
             "hostType": "npm",
             "matchHost": "https://private.repo.url/repository/npm/"
           }
         ]
```

This results in the bot ignoring my private registry and trying to fetch everything from the public npm registry.

I can see resetting npmrc log for every project in the given list
```
DEBUG: Resetting npmrc (repository=group1/project1)
DEBUG: Resetting npmrc (repository=group1/project2)
DEBUG: Resetting npmrc (repository=group2/project1)
```


## npm host rule + npmrc with only registry URL, token passed via hostRules
Same thing as the example above, token property is missing from hostRules in debug log. 
Every project tries to connect to private repo, but gets 401.

## npm host rule + npmrc with registry name and auth part (bare token provided in GitLab variable)
So, npmrc looks like `registry=https://private.repo.url/repository/npm/\n//private.repo.url/repository/npm/:authToken=${process.env.CI_NPM_REGISTRY_TOKEN}`
Again same as the first example, the first project will successfully connect to the private registry and pull dependencies, while 2nd and 3rd project gets 401. Again I can see debug log for setting npmrc for 1st project only
`
DEBUG: Setting npmrc (repository=group1/project1)
`


What I concluded is that:
- if npmrc property is present in the config, it will only be set for 1st project on the bot's to-do list.
- if token is provided inside host rule, it is omitted.? anyhow, all auth fails since the created npmrc file is faulty.

**Comment by rarkins** [comment]
If you ignore authentication for now and just set `npmrc` to `registry=https://private.repo.url/repository/npm/`, does that registry setting persist across projects, or only work for the first project? i.e. I expect it to *fail* on the first project due to lack of auth, but that's OK. The purpose is to reduce the example down to just the registry and test persistence.

**Comment by rarkins** [comment]
I would like to try "npm host rule + npmrc with only registry URL, token passed via hostRules"

If you say this option never works then it indicates there's a problem with the hostRules.

Please play around with this option, including setting a dummy string for token to see if *then* it gets printed/used

**Comment by kovasaurus** [comment, marked as answer]
You are onto something, when I provide a dummy string to token property inside host rules, I can see it being logged (but masked) on debug console. Seems like there is an issue with resolving it from the ENV variable. But I will play around a bit more and post results
